File: linjiashop-admin/src/main/java/com/ityu/linjiaadminapi/controller/AccountController.java
package com.ityu.linjiaadminapi.controller;


import com.ityu.bean.constant.state.ManagerStatus;
import com.ityu.bean.entity.system.User;
import com.ityu.bean.vo.front.Rets;
import com.ityu.cache.CacheDao;
import com.ityu.core.log.LogManager;
import com.ityu.core.log.LogTaskFactory;
import com.ityu.security.AdminUser;
import com.ityu.service.system.AccountService;
import com.ityu.service.system.ManagerService;
import com.ityu.utils.HttpUtil;
import com.ityu.utils.JwtTokenUtil;
import com.ityu.utils.Maps;
import com.ityu.utils.StringUtil;
import com.ityu.web.ApiConstants;
import com.ityu.web.controller.BaseController;
import org.nutz.mapl.Mapl;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.HashMap;
import java.util.Map;

/**
 * AccountController
 *
 * @author enilu
 * @version 2018/9/12 0012
 */
@RestController
@RequestMapping("/api/account")
public class AccountController extends BaseController {
    private Logger logger = LoggerFactory.getLogger(AccountController.class);



    @Autowired
    BCryptPasswordEncoder bCryptPasswordEncoder;


    @Value("${jwt.token.expire.time}")
    private Long tokenExpireTime;
    @Autowired
    private ManagerService managerService;
    @Autowired
    private AccountService accountService;

    @Autowired
    private CacheDao cacheDao;

    @RequestMapping(value = "/info", method = RequestMethod.GET)
    public Object info() {
        AdminUser shiroUser = getAdminUser();
        if (shiroUser == null) {
            return Rets.failure("请登录");
        }
        if (shiroUser.getRoleList().size() == 0) {
            return Rets.failure("未配置权限");
        }
        Map map = Maps.newHashMap("name", shiroUser.getName(), "role", "admin", "roles", shiroUser.getRoleCodes());
        map.put("permissions", shiroUser.getUrls());
        Map profile = (Map) Mapl.toMaplist(shiroUser);
        profile.put("dept", shiroUser.getDeptName());
        profile.put("roles", shiroUser.getRoleNames());
        map.put("profile", profile);
        return Rets.success(map);
    }

    @RequestMapping(value = "/updatePwd", method = RequestMethod.POST)
    public Object updatePwd(String oldPassword, String password, String rePassword) {
        try {
            if (StringUtil.isEmpty(password) || StringUtil.isEmpty(rePassword)) {
                return Rets.failure("密码不能为空");
            }
            if (!password.equals(rePassword)) {
                return Rets.failure("新密码前后不一致");
            }
            //获取用户信息
            User user = managerService.get(getAdminUser().getId());
            //不能修改超级管理员密码
            if (ApiConstants.ADMIN_ACCOUNT.equals(user.getAccount())) {
                return Rets.failure("不能修改超级管理员密码");
            }

            if (!bCryptPasswordEncoder.matches(oldPassword, user.getPassword())) {
                return Rets.failure("旧密码输入错误");
            }
            user.setPassword(bCryptPasswordEncoder.encode(password));
            managerService.update(user);
            //清空缓存
            cacheDao.hset(CacheDao.SESSION,user.getAccount(),null);
            return Rets.success();
        } catch (Exception e) {
            logger.error(e.getMessage(), e);
        }
        return Rets.failure("更改密码失败");
    }

    /**
     * 用户登录<br>
     * 1，验证没有注册<br>
     * 2，验证密码错误<br>
     * 3，登录成功
     *
     * @param mobile
     * @param password
     * @return
     */
    @RequestMapping(value = "/login", method = RequestMethod.POST)
    public Object login(@RequestParam("mobile") String mobile,
                        @RequestParam("password") String password) {
        try {
            logger.info("用户登录:" + mobile + ",密码:" + password);
            //1,
            User user = managerService.findByAccount(mobile);
            if (user == null) {
                return Rets.failure("该用户不存在");
            }
            if(user.getStatus() == ManagerStatus.FREEZED.getCode()){
                return Rets.failure("用户已冻结");
            }else if(user.getStatus() == ManagerStatus.DELETED.getCode()){
                return Rets.failure("用户已删除");
            }
            if (!bCryptPasswordEncoder.matches(password, user.getPassword())) {
                return Rets.failure("输入的密码错误");
            }
            String token = JwtTokenUtil.getInstance().generateToken(mobile,user.getId());
            String s = JwtTokenUtil.TOKEN_PREFIX + token;
            Map<String, String> result = new HashMap<>(1);
            logger.info("token:{}", s);
            result.put("token", s);
            return Rets.success(result);
        } catch (Exception e) {
            logger.error(e.getMessage(), e);
        }
        return Rets.failure("登录时失败");
    }

    /**
     * 退出登录
     *
     * @return
     */
    @RequestMapping(value = "/logout", method = RequestMethod.POST)
    public Object logout() {
        AdminUser shiroUser = getAdminUser();
        accountService.logout(shiroUser.getId());
        LogManager.me().executeLog(LogTaskFactory.exitLog(shiroUser.getId(), HttpUtil.getIp()));
        return Rets.success();
    }

}

File: sdf-api/src/main/java/com/ityu/sdf/api/controller/AccountController.kt
package com.ityu.sdf.api.controller

import com.ityu.common.bean.entity.system.User
import com.ityu.common.bean.vo.front.Ret
import com.ityu.common.bean.vo.front.Rets
import com.ityu.common.bean.vo.front.Rets.failure
import com.ityu.common.bean.vo.node.RouterMenu
import com.ityu.common.cache.CacheDao
import com.ityu.common.cache.TokenCache
import com.ityu.common.core.log.LogManager
import com.ityu.common.core.log.LogTaskFactory
import com.ityu.common.dto.LoginDto
import com.ityu.common.exception.Asserts
import com.ityu.common.exception.ResultCode
import com.ityu.common.service.system.MenuService
import com.ityu.common.service.system.QrcodeService
import com.ityu.common.service.system.UserService
import com.ityu.common.utils.*
import com.ityu.common.utils.HttpUtil.getToken
import com.ityu.sdf.api.bo.AdminUserDetails
import com.ityu.sdf.api.utils.ApiConstants
import io.swagger.annotations.Api
import io.swagger.annotations.ApiOperation
import io.swagger.v3.oas.annotations.security.SecurityRequirements
import io.swagger.v3.oas.annotations.tags.Tag
import org.slf4j.LoggerFactory
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.web.bind.annotation.*

/**
 * AccountController
 *
 * @author enilu
 * @version 2018/9/12 0012
 */
@Api(tags = ["account"])
@Tag(name = "account", description = "账户模块")
@RestController
@RequestMapping("/account")
open class AccountController(
    private var userService: UserService,
    private var menuService: MenuService,
    private var passwordEncoder: PasswordEncoder,
    private var jwtTokenUtil: JwtTokenUtil,
    private var tokenCache: TokenCache,
    private var cacheDao: CacheDao,
    private var qrcodeService: QrcodeService,
) {
    private val logger = LoggerFactory.getLogger(AccountController::class.java)

    /**
     * 用户登录<br></br>
     * 1，验证没有注册<br></br>
     * 2，验证密码错误<br></br>
     * 3，登录成功
     *
     * @param loginDto
     * @return
     */
    @PostMapping(value = ["/login"])
    @SecurityRequirements
    fun login(@RequestBody loginDto: LoginDto?): Ret<*> {
        var token: String?
        //密码需要客户端加密后传递
        //密码需要客户端加密后传递
        try {
            val result = userService.loadUserByUsername(loginDto!!.username)
            val userDetails = AdminUserDetails(result.first, result.second)
            val password = CryptUtil.desEncrypt(loginDto.password)
            if (!passwordEncoder.matches(password, userDetails.password)) {
                Asserts.fail("密码不正确")
            }
            if (!userDetails.isEnabled) {
                Asserts.fail("帐号已被禁用")
            }
            val authentication = UsernamePasswordAuthenticationToken(userDetails, null, userDetails.authorities)
            SecurityContextHolder.getContext().authentication = authentication
            token = jwtTokenUtil.generateToken(userDetails.username, result.first.id)
            LogManager.me().executeLog(LogTaskFactory.loginLog(userDetails.umsAdmin.id, HttpUtil.getIp()))
            val map = mapOf(
                "token" to token
            )
            userService.setAdminUser(token, result.first)
            return Rets.success(map)
        } catch (e: AuthenticationException) {
            logger.warn("登录异常:{}", e.message)
        }
        return failure("登录时失败", null)
    }

    @GetMapping(value = ["/info"])
    fun info(): Any {
        var id: Long? = null
        try {
            id = HttpUtil.getUserId(jwtTokenUtil)
        } catch (e: Exception) {
            return Rets.expire()
        }
        if (id != null) {
            var user: User? = userService.get(id) ?: //该用户可能被删除
            return Rets.expire()
            if (StringUtil.isEmpty(user!!.roleid)) {
                return failure("该用户未配置权限")
            }
            val shiroUser = tokenCache.getUser(jwtTokenUtil.getTokenRemoveHead(getToken()))

            if (shiroUser == null) {
                Asserts.fail(ResultCode.UNAUTHORIZED)
            }
            val list: List<RouterMenu> = menuService.getSideBarMenus(shiroUser!!.roleList)

            val result = mapOf(
                "roles" to shiroUser!!.roleCodes,
                "username" to user.account,
                "name" to user.name,
                "menus" to list,
                "permissions" to shiroUser.urls,
                "profile" to mapOf(
                    "dept" to shiroUser.deptName,
                    "roles" to shiroUser.roleNames,
                ),
            )
            return Rets.success(result)
        }
        return failure("获取用户信息失败")
    }

    @PostMapping(value = ["/updatePwd"])
    fun updatePwd(oldPassword: String?, password: String, rePassword: String): Any {
        try {
            if (StringUtil.isEmpty(password) || StringUtil.isEmpty(rePassword)) {
                return failure("密码不能为空")
            }
            if (password != rePassword) {
                return failure("新密码前后不一致")
            }
            val user: User? = userService.findByAccountForLogin(HttpUtil.getUsername(jwtTokenUtil))
            if (ApiConstants.ADMIN_ACCOUNT == user?.account) {
                return failure("不能修改超级管理员密码")
            }
            val rawPassword = CryptUtil.desEncrypt(oldPassword)
            if (!passwordEncoder.matches(rawPassword, user?.password)) {
                Asserts.fail("旧密码输入错误")
            }

            user?.password = passwordEncoder.encode(password)
            userService.update(user)
            return Rets.success()
        } catch (e: java.lang.Exception) {
            logger.error(e.message, e)
        }
        return failure("更改密码失败")
    }

    /**
     * @param phone 用户账号
     * @param qrcode  二维码值
     * @param confirm 是否确认登录：1:是,0:否
     * @return
     */
    @PostMapping("/qrcode/login")
    fun qrLogin(
        @RequestParam("phone") phone: String?,
        @RequestParam("qrcode") qrcode: String?,
        @RequestParam("confirm") confirm: String?
    ): Ret<*> {
        val qrstatus: String = qrcodeService.getCrcodeStatus(qrcode)
        if (QrcodeService.INVALID == qrstatus) {
            return failure("二维码已过期")
        }
        if (QrcodeService.SUCCESS == qrstatus || QrcodeService.CANCEL == qrstatus) {
            return failure("二维码已被他人使用")
        }
        return if (QrcodeService.UNDO == qrstatus) {
            if ("0" == confirm) {
                cacheDao.hset(CacheDao.SHORT, qrcode, QrcodeService.CANCEL)
                failure("已取消")
            }
            val result = userService.loadUserByPhone(phone)
            val userDetails = AdminUserDetails(result.first, result.second)
            val authentication = UsernamePasswordAuthenticationToken(userDetails, null, userDetails.authorities)
            SecurityContextHolder.getContext().authentication = authentication

            val token = userService.loginForToken(result.first)

            userService.setAdminUser(token, result.first)
            cacheDao.hset(
                CacheDao.SHORT, qrcode, JsonUtil.toJson(
                    Maps.newHashMap(
                        "status", QrcodeService.SUCCESS, "account", result.first.account, "token", token
                    )
                )
            )
            Rets.success()
        } else if (QrcodeService.INVALID == qrstatus) {
            failure("二维码已过期")
        } else {
            failure("无效的二维码")
        }
    }

    /**
     * 获取二维码扫码结果(PC端调用）
     *
     * @return
     */
    @GetMapping("/qrcode/getRet")
    fun getQrcodeStatus(@RequestParam("uuid") uuid: String?): Ret<*> {
        val ret = qrcodeService.getCrcodeStatus(uuid)
        if (QrcodeService.INVALID == ret) {
            return Rets.success(Maps.newHashMap("status", ret, "msg", "二维码已过期"))
        }
        if (QrcodeService.CANCEL == ret) {
            return Rets.success(Maps.newHashMap("status", ret, "msg", "已取消登录"))
        }
        if (QrcodeService.UNDO == ret) {
            return Rets.success(Maps.newHashMap("status", ret, "msg", "待扫描"))
        }
        val map: Map<*, *> = JsonUtil.fromJson(MutableMap::class.java, ret)
        return Rets.success(map)
    }


    @ApiOperation(value = "登出功能")
    @RequestMapping(value = ["/logout"], method = [RequestMethod.POST])
    @ResponseBody
    fun logout(): Ret<*> {
        tokenCache.remove(getToken())
        return Rets.success(null)
    }

}
